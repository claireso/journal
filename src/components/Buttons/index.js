import PropTypes from 'prop-types'
import * as S from './Buttons.styles'

export const SecondaryButton = S.SecondaryButton
export const ButtonIcon = S.ButtonIcon
export const StyledButton = S.StyledButton

export const PrimaryButton = ({ isLoading, ...props }) => {
  return isLoading ? <S.PrimaryButtonLoading /> : <S.PrimaryButton {...props} />
}

PrimaryButton.propTypes = {
  isLoading: PropTypes.bool
}
