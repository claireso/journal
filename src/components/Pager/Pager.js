import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import * as S from './Pager.styles'

const Pager = ({ navigate, first, prev, next, last }) => {
  const handleClick = useCallback((page) => () => navigate(page), [navigate])

  return (
    <S.Wrapper>
      {first && (
        <li>
          <S.Button data-testid="first-page" title="First page" onClick={handleClick(first)}>
            ««
          </S.Button>
        </li>
      )}

      {prev && (
        <li>
          <S.Button data-testid="previous-page" title="Previous page" onClick={handleClick(prev)}>
            «
          </S.Button>
        </li>
      )}

      {next && (
        <li>
          <S.Button data-testid="next-page" title="Next page" onClick={handleClick(next)}>
            »
          </S.Button>
        </li>
      )}

      {last && (
        <li>
          <S.Button data-testid="last-page" title="Last page" onClick={handleClick(last)}>
            »
          </S.Button>
        </li>
      )}
    </S.Wrapper>
  )
}

Pager.propTypes = {
  navigate: PropTypes.func.isRequired,
  first: PropTypes.number,
  prev: PropTypes.number,
  next: PropTypes.number,
  last: PropTypes.number
}

export default Pager
