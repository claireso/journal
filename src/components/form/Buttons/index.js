export { default } from './Button'
export { SubmitButton } from './Buttons.styles'
