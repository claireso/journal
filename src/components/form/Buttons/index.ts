export { default } from './Button'
export { ButtonSubmit } from './Buttons.styles'
