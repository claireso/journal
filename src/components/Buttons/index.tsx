import * as S from './Buttons.styles'
import { Loader } from '../Loader'

interface ButtonPrimaryProps {
  children: React.ReactNode
  isLoading?: boolean
  onClick?: () => void
  title?: string
}

export const ButtonPrimary = ({ isLoading = false, ...props }: ButtonPrimaryProps) => {
  if (isLoading) {
    return (
      <S.ButtonLoading color="primary">
        <Loader />
      </S.ButtonLoading>
    )
  }
  return <S.Button color="primary" {...props} />
}

export const ButtonSecondary = (props: Omit<ButtonPrimaryProps, 'isLoading'>) => (
  <S.Button color="secondary" {...props} />
)
export const ButtonIcon = S.ButtonIcon
export const Button = S.Button
