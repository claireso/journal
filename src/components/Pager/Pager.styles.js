import styled from 'styled-components'

import { SecondaryButton } from '../Buttons'

export const PagerWrapper = styled.ul`
  display: inline-flex;
  justify-content: center;
  list-style-type: none;
  width: 100%;
  padding: 0;
  > li {
    margin: 0 0.4rem;
  }
`

export const PagerButton = styled(SecondaryButton)`
  align-items: center;
  font-size: var(--font-size-normal);
  height: 4.4rem;
  justify-content: center;
  padding: 0;
  width: 4.4rem;
`
