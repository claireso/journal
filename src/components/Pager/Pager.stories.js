import React from 'react'
import { action } from '@storybook/addon-actions'

import Pager from './index'

export default {
  title: 'Pager'
}

export const Basic = () => {
  const data = {
    first: 1,
    last: 8,
    next: 3,
    prev: 1
  }

  return <Pager {...data} navigate={action('on page change')} />
}
