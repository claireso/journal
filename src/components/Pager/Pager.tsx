import React, { useCallback } from 'react'

import * as S from './Pager.styles'

interface PagerProps {
  navigate: (page: string) => void
  first?: number
  prev?: number
  next?: number
  last?: number
}

const Pager = ({ navigate, first, prev, next, last }: PagerProps) => {
  const handleClick = useCallback((page) => () => navigate(`${page}`), [navigate])

  return (
    <S.Wrapper>
      {first && (
        <li>
          <S.Button data-testid="first-page" title="First page" onClick={handleClick(first)}>
            ««
          </S.Button>
        </li>
      )}

      {prev && (
        <li>
          <S.Button data-testid="previous-page" title="Previous page" onClick={handleClick(prev)}>
            «
          </S.Button>
        </li>
      )}

      {next && (
        <li>
          <S.Button data-testid="next-page" title="Next page" onClick={handleClick(next)}>
            »
          </S.Button>
        </li>
      )}

      {last && (
        <li>
          <S.Button data-testid="last-page" title="Last page" onClick={handleClick(last)}>
            »
          </S.Button>
        </li>
      )}
    </S.Wrapper>
  )
}

export default Pager
