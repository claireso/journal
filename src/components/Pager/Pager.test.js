import { render, fireEvent } from '@testing-library/react'

import Pager from './index'
import { PagerButton } from '../Buttons'

const getPager = (props = {}) => (
  <Pager {...props}>
    {({ items, getItemsProps }) =>
      items.map((item, key) => (
        <li key={key}>
          <PagerButton
            {...getItemsProps({
              label: item.label,
              title: item.title,
              item: item
            })}
          />
        </li>
      ))
    }
  </Pager>
)

describe('<Pager />', () => {
  test('should not render items', () => {
    const { container } = render(getPager({ navigate: () => {} }))

    expect(container.querySelector('li')).toBeNull()
  })

  test('should render first page', () => {
    const navigate = jest.fn()
    const { container, getByTitle } = render(
      getPager({
        next: global.__PHOTOS__.pager.next,
        last: global.__PHOTOS__.pager.last,
        navigate: navigate
      })
    )

    expect(container.querySelectorAll('li')).toHaveLength(2)

    expect(getByTitle('Next page')).toBeInTheDocument()
    fireEvent.click(getByTitle('Next page'))
    expect(navigate).toHaveBeenCalledWith(2)

    expect(getByTitle('Last page')).toBeInTheDocument()
    fireEvent.click(getByTitle('Last page'))
    expect(navigate).toHaveBeenCalledWith(19)
  })

  test('should render second page', () => {
    const navigate = jest.fn()
    const { container, getByTitle } = render(
      getPager({
        first: 1,
        prev: 1,
        next: global.__PHOTOS__.pager.next,
        last: global.__PHOTOS__.pager.last,
        navigate: navigate
      })
    )

    expect(container.querySelectorAll('li')).toHaveLength(4)

    expect(getByTitle('First page')).toBeInTheDocument()
    fireEvent.click(getByTitle('First page'))
    expect(navigate).toHaveBeenCalledWith(1)

    expect(getByTitle('Previous page')).toBeInTheDocument()
    fireEvent.click(getByTitle('Previous page'))
    expect(navigate).toHaveBeenCalledWith(1)

    expect(getByTitle('Next page')).toBeInTheDocument()
    fireEvent.click(getByTitle('Next page'))
    expect(navigate).toHaveBeenCalledWith(2)

    expect(getByTitle('Last page')).toBeInTheDocument()
    fireEvent.click(getByTitle('Last page'))
    expect(navigate).toHaveBeenCalledWith(19)
  })

  test('should render last page', () => {
    const navigate = jest.fn()
    const { container, getByTitle } = render(
      getPager({
        first: 1,
        prev: 1,
        navigate: navigate
      })
    )

    expect(container.querySelectorAll('li')).toHaveLength(2)

    expect(getByTitle('First page')).toBeInTheDocument()
    fireEvent.click(getByTitle('First page'))
    expect(navigate).toHaveBeenCalledWith(1)

    expect(getByTitle('Previous page')).toBeInTheDocument()
    fireEvent.click(getByTitle('Previous page'))
    expect(navigate).toHaveBeenCalledWith(1)
  })
})
