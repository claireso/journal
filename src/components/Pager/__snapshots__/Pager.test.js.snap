// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pager /> should not render items 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  width: 100%;
  padding: 0;
}

.c0 > li {
  margin: 0 0.4rem;
}

<ul
    class="c0"
  />
</DocumentFragment>
`;

exports[`<Pager /> should render first page 1`] = `
<DocumentFragment>
  .c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: var(--font-family);
  outline: none;
  padding: 0;
  margin: 0;
}

.c3 {
  border-radius: 0.2rem;
  font-size: var(--font-size-smaller);
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
}

.c3 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c1 + .c1 {
  margin: 0 0 0 0.8rem;
}

.c4 {
  background: var(--secondary-normal);
  color: var(--text-normal);
}

.c4:hover {
  background: var(--secondary-darker);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  width: 100%;
  padding: 0;
}

.c0 > li {
  margin: 0 0.4rem;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: var(--font-size-normal);
  height: 4.4rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
  width: 4.4rem;
}

<ul
    class="c0"
  >
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="next-page"
        title="Next page"
      >
        »
      </button>
    </li>
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="last-page"
        title="Last page"
      >
        »
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<Pager /> should render last page 1`] = `
<DocumentFragment>
  .c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: var(--font-family);
  outline: none;
  padding: 0;
  margin: 0;
}

.c3 {
  border-radius: 0.2rem;
  font-size: var(--font-size-smaller);
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
}

.c3 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c1 + .c1 {
  margin: 0 0 0 0.8rem;
}

.c4 {
  background: var(--secondary-normal);
  color: var(--text-normal);
}

.c4:hover {
  background: var(--secondary-darker);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  width: 100%;
  padding: 0;
}

.c0 > li {
  margin: 0 0.4rem;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: var(--font-size-normal);
  height: 4.4rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
  width: 4.4rem;
}

<ul
    class="c0"
  >
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="first-page"
        title="First page"
      >
        ««
      </button>
    </li>
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="previous-page"
        title="Previous page"
      >
        «
      </button>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<Pager /> should render second page 1`] = `
<DocumentFragment>
  .c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: var(--font-family);
  outline: none;
  padding: 0;
  margin: 0;
}

.c3 {
  border-radius: 0.2rem;
  font-size: var(--font-size-smaller);
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
}

.c3 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c1 + .c1 {
  margin: 0 0 0 0.8rem;
}

.c4 {
  background: var(--secondary-normal);
  color: var(--text-normal);
}

.c4:hover {
  background: var(--secondary-darker);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  width: 100%;
  padding: 0;
}

.c0 > li {
  margin: 0 0.4rem;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: var(--font-size-normal);
  height: 4.4rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
  width: 4.4rem;
}

<ul
    class="c0"
  >
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="first-page"
        title="First page"
      >
        ««
      </button>
    </li>
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="previous-page"
        title="Previous page"
      >
        «
      </button>
    </li>
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="next-page"
        title="Next page"
      >
        »
      </button>
    </li>
    <li>
      <button
        class="c1 c2 c3 c4 c5"
        data-testid="last-page"
        title="Last page"
      >
        »
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
